base/inference.jl

```diff
@@ -2363,15 +2363,77 @@ end
 
 #### post-inference optimizations ####
 
-function inline_as_constant(val::ANY, argexprs, sv::InferenceState)
+immutable InvokeData
+    mt::MethodTable
+    entry::TypeMapEntry
+    types0
+    fexpr
+    texpr
+    orig_atypes
+end
+
+const invoke_arg_error = ErrorException("invoke: argument type error")
+
+# Assume `argexprs` that we need are all `effect_free`
+function emit_invoke_typecheck!(argexprs, invoke_data::InvokeData, stmts, sv)
```

Comment:
I think this function is doing too much work --- it's an ad-hoc implementation of `isa` for tuple types. Ideally it would do a single call to `isa`, or otherwise ensure that the same exact check is done as in normal dispatch. For example, soon `invoke_data.types0` might be a UnionAll type instead of a Tuple type.